sdks/csharp/sdk/Alpha.cs

```diff
@@ -210,6 +211,380 @@ public async Task<List<string>> GetConnectedPlayersAsync()
+        {
+          try
+          {
+                var request = new GetCounterRequest()
+                {
+                    Name = key,
+                };
+                var counter = await client.GetCounterAsync(request,
+              deadline: DateTime.UtcNow.AddSeconds(RequestTimeoutSec), cancellationToken: ctoken);
+            return counter.Count;
+          }
+          catch (RpcException ex)
+          {
+                LogError(ex, $"Unable to invoke GetCounterCount({key}).");
+                throw;
+            }
+        }
+
+        /// <summary>
+        /// IncrementCounterAsync increases a counter by the given nonnegative integer amount.
+        /// Will execute the increment operation against the current CRD value. Will max at max(int64).
+        /// Will error if the key was not predefined in the GameServer resource on creation.
+        /// Returns false if the count is at the current capacity (to the latest knowledge of the SDK),
+        /// and no increment will occur.
+        ///
+        /// Note: A potential race condition here is that if count values are set from both the SDK and
+        /// through the K8s API (Allocation or otherwise), since the SDK append operation back to the CRD
+        /// value is batched asynchronous any value incremented past the capacity will be silently truncated.
+        /// </summary>
+        /// <returns>True if the increment counter request was successful.</returns>
+        public async Task<bool> IncrementCounterAsync(string key, long amount)
+        {
+            if (amount < 0)
+            {
+                throw new ArgumentOutOfRangeException($"CountIncrement amount must be a positive number, found {amount}");
+            }
+            try
+            {
+                var request = new CounterUpdateRequest()
+                {
+                    Name = key,
+                    CountDiff = amount,
+                };
+                var updateRequest = new UpdateCounterRequest()
+                {
+                    CounterUpdateRequest = request,
+                };
+                var response = await client.UpdateCounterAsync(updateRequest,
+                  deadline: DateTime.UtcNow.AddSeconds(RequestTimeoutSec), cancellationToken: ctoken);
+                // If we get a response (Counter) without an error, then the request was successful.
+                return true;
+            }
+            catch (RpcException ex)
+            {
+                LogError(ex, $"Unable to invoke IncrementCounter({key}, {amount}).");
+                throw;
+            }
+        }
+
+        /// <summary>
+        /// DecrementCounterAsync decreases the current count by the given nonnegative integer amount.
+        /// The Counter Will not go below 0. Will execute the decrement operation against the current CRD value.
+        /// Returns false if the count is at 0 (to the latest knowledge of the SDK), and no decrement will occur.

```

Comment:
`Returns false if the count is at 0`

This method always returns true, should it be returning a value from the response object?

Reply:
This seems to be the case for other bool-returning methods in this file, they may all need updates?

Reply:
This is part of the weirdness of the API design #2716 expecting a `bool` return value, which doesn't match with the standard gRPC calls https://github.com/googleforgames/agones/blob/929a3e71c68d3992ba50a96f251c3047fe523b18/sdks/csharp/proto/sdk/alpha/alpha.proto#L138-L146 that take in a `UpdateCounterRequest` and return a `Counter` (similar pattern for `List` as well). The SDK server follows the same pattern of taking in an update request and return the updated object https://github.com/googleforgames/agones/blob/929a3e71c68d3992ba50a96f251c3047fe523b18/pkg/sdkserver/sdkserver.go#L849. Returning `true` if the SDK returns an object, and throwing an exception if the SDK returns an error, is how I tried to work around this. Do you have a suggestion for a better way to handle this?

Reply:
Interesting...honestly it doesn't seem like this SDK is that useful, for C# at least. I notice that design doc says `The SDK will batch operations every 1 second for performance reasons, but changes made through the SDK will be atomically accurate through the SDK.` but unless gRPC is doing this internally, this SDK doesn't seem to do anything of the sort.

It's just a super thin wrapper on top of a gRPC client, and at this point it seems more useful to just use the generated gRPC client itself? I suppose the SDK is helping to craft various requests here, but other than that (which already shouldn't be difficult), it's not doing much.

If this design is kept, I would expect the documentation here is at least updated to reflect the actual implementation, or the implementation updated to match the documentation, whichever is appropriate. It seems like the design doc is expecting the SDK to be doing a lot more than it currently is.

Reply:
> It's just a super thin wrapper on top of a gRPC client

all the SDKs are thin wrappers over a gRPC client. Or thin wrappers over a HTTP client. The meat of the logic is stored in the sdk server. Hence we can have multiple language SDKs without the overhead or repeating a lot of logic.

> but unless gRPC is doing this internally, this SDK doesn't seem to do anything of the sort.

That's exactly what the SDK server does. The SDK Server plus the language library are "the SDK" as a whole.

>If this design is kept, I would expect the documentation here is at least updated to reflect the actual implementation, or the implementation updated to match the documentation, whichever is appropriate. It seems like the design doc is expecting the SDK to be doing a lot more than it currently is.

Not sure I follow here. The documentation and design follow what the functionality of the SDK does - how the implementation works should be fairly immaterial to the end user.

> It seems like the design doc is expecting the SDK to be doing a lot more than it currently is.

I wrote the design doc 😄 this is doing exactly what it is supposed to.

Reply:
My bad, I was asked to take a quick look at this and being new to the project, I obviously missed some core architectural pieces 😅 

I suppose my feedback boils down to this: when/how should this function return false? If it never does, perhaps it should change to a void (`Task`) function?

Reply:
Would it make sense to, instead of throwing an error, to log the error and return false? Or is the void (Task) function more idiomatic in C#? (And for my new-to-C# understanding, are you recommending `public async Task<void> DecrementCounterAsync(string key, long amount)`?)

Reply:
> I suppose my feedback boils down to this: when/how should this function return false? If it never does, perhaps it should change to a void (`Task`) function?

"Returns false if the count is at 0 (to the latest knowledge of the SDK), and no decrement will occur."

Basically a false is a "hey, this is a no-op, just letting you know"- so to this point, the SDK shouldn't be throwing an exception here.

Reply:
Digging into this a bit more - the original design of this function was such that if the counter was at capacity (0,capacity), then it should return false, not that it should error -- whereas now it throws an exception.

@igooch and I were talking about this offline, and would love your feedback on what the appropriate response here should be - @Jensaarai and @castaneai (since we have you both here), what would you prefer to happen here if you try to decrement below 0, or increment above capacity?

Reply:
My feeling is that throwing an exception (more specifically, InvalidOperationException) is more appropriate as C# API.

Reply:
@igooch I believe you were also advocating for an exception return as well? @Jensaarai how do you feel?

I was arguing  for the boolean, but happy to be out-voted if y'all are in agreement 😃 - if we are  drawing consensus to throwing an exception here, then I would also agree to returning `void` from the function, and drop the boolean entirely (and discuss probably doing this for all other SDKs as well).

Reply:
I'm not really familiar with Go, but it appears that its `error` is roughly equivalent to an exception in C#? https://go.dev/blog/error-handling-and-go

In that case, it seems like these functions shouldn't return a bool, and in Go a non-nil error would denote failure, while in C# it would be an exception.

What I'm not sure about is the behavior of the Go gRPC client vs. C#. For instance, here in Go, there's a call to `UpdateCounter` that can return `err`, and if it does, it gets wrapped and returned:
```Go
// DecrementCounter decreases the current count by the given nonnegative integer amount.
// The Counter Will not go below 0. Will execute the decrement operation against the current CRD value.
// Returns false if the count is at 0 (to the latest knowledge of the SDK), and no decrement will occur.
func (a *Alpha) DecrementCounter(key string, amount int64) (bool, error) {
	if amount < 0 {
		return false, errors.Errorf("CountDecrement amount must be a positive int64, found %d", amount)
	}
	_, err := a.client.UpdateCounter(context.Background(), &alpha.UpdateCounterRequest{
		CounterUpdateRequest: &alpha.CounterUpdateRequest{
			Name:      key,
			CountDiff: amount * -1,
		}})
	if err != nil {
		return false, errors.Wrapf(err, "could not decrement Counter %s by amount %d", key, amount)
	}
	return true, err
}
```

In C#, we have:
```csharp
        public async Task<bool> DecrementCounterAsync(string key, long amount)
        {
            if (amount < 0)
            {
                throw new ArgumentOutOfRangeException($"DecrementCounter amount must be a positive number, found {amount}");
            }
            try
            {
                var request = new CounterUpdateRequest()
                {
                    Name = key,
                    CountDiff = amount * -1,
                };
                var updateRequest = new UpdateCounterRequest()
                {
                    CounterUpdateRequest = request,
                };
                var response = await client.UpdateCounterAsync(updateRequest,
                  deadline: DateTime.UtcNow.AddSeconds(RequestTimeoutSec), cancellationToken: ctoken);
                return true;
            }
            catch (RpcException ex)
            {
                LogError(ex, $"Unable to invoke DecrementCounter({key}, {amount}).");
                throw;
            }
        }
```

I imagine that here, the `UpdateCounterAsync` gRPC method would throw an `RpcException` in the same `err` case that Go is handling? In which case, I think the re-throw is already appropriate, and thus the boolean is not that useful.

Additionally, I think the `ArgumentOutOfRangeException` for `amount < 0` is the appropriate thing for C# to do :)

Reply:
Is the recommended action here to update `public async Task<bool> DecrementCounterAsync(string key, long amount)` to `public async Task DecrementCounterAsync(string key, long amount)`? If it is, @Jensaarai @castaneai do you have recommendations for how to update the tests? (I'm running into issues like ` error CS0029: Cannot implicitly convert type 'void' to 'System.Threading.Tasks.Task'`.)

Reply:
Yes. Changing to `Task` is correct. In the test, you don't need to use response, only await.

Reply:
@igooch would you like me to write up a ticket to update the SDK methods globally to remove the boolean from count increment/decrement operations?

Reply:
Yes, @markmandel please write up a ticket to update the bool from the operations that don't use it. I removed it from the IncrementCounterAsync, DecrementCounterAsync, SetCounterCountAsync, SetCounterCapacityAsync, SetListCapacityAsync, AppendListValueAsync, and DeleteListValueAsync.

Reply:
@castaneai @Jensaarai the change from `public async Task<bool>` to `public async Task` is ready for review when you have the time.

Reply:
@castaneai did you want to review one last time before we merge the PR?